Fix Block Editor Iframe component to render in standards mode

tomasztunik opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
This fixes an issue where iframe component used by Block Editor would 
render its contents differently to how they were rendered on the front-end.

Iframe document doctype is not set by default if it is not provided via iframe 
`src` or `srcDoc` and not having doctype will cause browser to render iframe 
contents in quirks mode. Appending doctype to existing document won't change 
iframe's rendering mode. `Document.write` will overwrite the document with 
a new one containing a correct doctype and will correctly render in stadards mode.

I've reviewed other areas of Gutenberg and it seems this is the only place
affected. All other places are correctly setting `<!DOCTYPE html>`.

More details provided in the related issue.

Fixes #38854

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->

1. Go to Template or Template Part editor.
2. Add block or edit template part containing quirks mode affected CSS/HTML. (Ie. [Mini Cart](https://github.com/woocommerce/woocommerce-gutenberg-products-block))
3. Confirm that contents render the same in the editor and in the front-end

## Screenshots <!-- if applicable -->




#### Example error:
|css|preview|
|-|-|
| ![table-quirks-mode](https://user-images.githubusercontent.com/112270/154259940-c4c903c2-01e2-4ad2-b53e-318c804234c8.jpg) | ![quirks-mode](https://user-images.githubusercontent.com/112270/154255814-141ee688-30f6-410e-86df-b194852fe184.jpg) |

#### Example valid after applying fix:
|css|preview|
|-|-|
| ![table-standards-mode](https://user-images.githubusercontent.com/112270/154259961-292eb10d-8018-4b12-ab92-3e91f44cd8e9.jpg) | ![standard-mode](https://user-images.githubusercontent.com/112270/154255829-9a14d840-13e7-4464-b53c-c8d6a9670504.jpg) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

tomasztunik committed:
fix Block Editor Iframe component to render in standards mode

Iframe document doctype is not set by default and not having doctype
will cause browser to render iframe contents in quirks mode. Appending
doctype to existing document won't change iframe's rendering mode.
Document.write will overwrite the document with a new one containing
a correct doctype and will correctly render in stadard mode.

Fixes #38854

ellatrix commented:
@tomasztunik Did you find any resources about this issue? Is there no other way to achieve it? I tried to use `document.write` at some point and it caused the contents to load quite a bit slower, so that's worth comparing. Although if there's styling issue, we don't really have much choice. We also need to test in other browsers to confirm that the iframe works, we only have test coverage for Chrome.

tomasztunik commented:
👋  @ellatrix - thanks for looking into this. 

As `document.write` happens only once on Iframe setup while the iframe document is still empty I did not notice any performance degradation on my end. Tried to run performance profiling on impact but couldn't really notice difference - might be dev machine is just too much for it to make it significant.

Were your findings around using document.write around iframe as well or were they part of DOM rendering as part of page loading flow?

Before I made it to this solution I've tried to populate the initial state using `srcDoc`, but this conflicted with different areas as it seemed to happen async with react ref setup hook. After that I've learned the hard way that appending doctype to already rendered document doesn't update the document's rendering mode if it was already rendered. So if I tried to use `document.implementation.createDocumentType('html', '', '')` and prepending it to the document even though you'd be able to see it in inspector it would still remain rendered in quirks mode.

Confirmed the bug to be present on latest Safari (16.1) and Firefox (97) and that the fix works there as well.

tomasztunik commented:
PS. re links - I've shared some resources on quirks mode on the issue (but I guess they were bit hidden in the description!):
- https://quirks.spec.whatwg.org/
- https://html.spec.whatwg.org/multipage/rendering.html (page search "In quirks mode")
- https://www.quirksmode.org/css/quirksmode.html

and additionally:
- https://html.spec.whatwg.org/multipage/syntax.html#the-doctype

ellatrix commented:
Would it be possible to add an e2e test with some example CSS, then check the result with `getComputedStyle`?

tomasztunik committed:
Tests validating site editor's iframe rendering mode

it will visit the editor and ensure that iframe is rendering
in standards CSS1Compat rendering mode. If it had
BackCompat set it would render in quirks mode.

tomasztunik commented:
I've added E2E tests that validate that site editor iframe is running in a correct `compatMode` ([ref](https://developer.mozilla.org/en-US/docs/Web/API/Document/compatMode)). I decided to go with this approach as I thought quirks mode can have different impact as mentioned in the docs linked earlier, and some of its side effects could be masked with css reset etc. This way we'll be sure even if all quirks were masked :) 

Tested to confirm that prior to code change it would fail the test with `BackCompat` value set.

Also tried to do it as unit tests with testing-libarary or enzyme rendering but it's super though to test iframes! 😵

(Noticed the tests have failed but it was some kind of timeout on one of the instances in unrelated area of the code)

tomasztunik committed:
restart CI after flaky test failure

jsnajdr commented:
Which service generates the actual document inside the iframe? And can we fix that service to send a correct doctype in the first place?

If we replace the document with a new one (doctype + original) when `readyState` becomes `interactive` or `complete`, doesn't that mean we'll load the document twice? That includes running all the inline scripts again, loading all the resources etc.

tomasztunik commented:
@jsnajdr by default empty iframe has no doctype (renders in quirks mode). And it's the responsibility of content provider to set it up. It can be done either via src (and it would come from remote content provider) or srcDoc if content provider is local. In theory it should be possible to setup initial document state - including doctype - via srcDoc as mentioned but this caused errors in unrelated areas because of how iframe control is handed over to react via React.Portal. Made some attempts toward that but the surface of the change started getting out of hand and would result in a much more complex code. 

And regarding your concerns, this wouldn't cause the contents to setup twice as this happens before control is given back to react and before anything is rendered into the iframe, including scripts and styles as the iframe document override is synchronous and happens before these are processed. 

Was trying to cpu profile that change prompted by @ellatrix but it wasn't even showing up in the logs and perceptive rendering speed doesn't seem to be affected.

jsnajdr commented:
Looking at the `Iframe` component that's being modified, the content of the `iframe` is managed by React, which renders a React UI into it, using a portal. It doesn't really support a `src` attribute, because it would overwrite the iframe markup with its own React UI?

So the task to do is to setup the initial empty document with a correct `<!DOCTYPE html>` tag? Yeah, then it makes sense, although instead of `document.write` we could write just `document.innerHTML = '<!DOCTYPE html>'` or init the iframe from a data URL:
```
<iframe src="data:<!DOCTYPE html>"/>
```

> Appending doctype to existing document won't change iframe's rendering mode.

This sounds like the `document.innerHTML` solution won't work and we really need to `document.write`?

> I've tried to populate the initial state using srcDoc, but this conflicted with different areas as it seemed to happen async with react ref setup hook.

`srcDoc` would be preferrable to `document.write`, and seems very similar to `<iframe src="data:..."/>`. Yes, it's probably async because `document.srcDoc = '...'` doesn't load immediately, but behaves as if it was loaded from a "virtual URL". We'd need to wait for some load event before calling `setIframeDocument` and triggering React render into the portal. But we already do that, because the rendering happens inside `setDocumentIfReady`?

Where were you calling `document.srcDoc =` when testing? The `load` event is probably too late.

tomasztunik commented:
Modifying existing document was my first instinct as well as mentioned [here](https://github.com/WordPress/gutenberg/pull/38855#issuecomment-1041577123) but unfortunately once document rendering mode is set appending doctype doesn't change it despite doctype being visible in DOM.

I was passing static `srcDoc` as prop to the `iframe` element with correct doctype included but as you mention - this caused initial empty document to load and then be asynchronously replaced with the one provided by srcDoc - you've described it better than I did before I think - so document replace happened regardless. This would force us to add complexity in waiting for iframe to be ready by having to skip first initialisation with wrong `compatMode` and wait for the correct one to be loaded. This replace was causing React which already took over content to act as if someone pulled carpet from under its legs.

The way proposed here seemed cleaner to me from code complexity as it was doing the same just made it synchronous within the iframe binding/setup method as we were already giving the control of the iframe to react and felt like this will keep the concern encapsulated in this method.

If you think this is not the case I could try to look again at the `srcDoc` approach. Now that we talk about that I'm thinking if waiting for right compatMode rather than interactive state wouldn't be enough... Let me think if we should give it a go :)

jsnajdr commented:
If you initialize the iframe as `<iframe srcDoc="..."/>` or `<iframe src="data:...">` then it should load that document right away, without loading an empty (`about:blank`) document first. I think it's worth trying one more time, and if it doesn't work, trying to observe carefully what exactly does not work.

Using `document.write` is acceptable, I just would write it as:
```js
document.open();
document.write( '<!DOCTYPE html>' );
document.close();
```
I think this is going to be a little less efficient because the API is synchronous -- like other legacy things like synchronous XHR requests or `window.alert`. But other than that, it's fine.

The `Iframe` component should be also modified to not accept a `src` prop, because from what I see it won't work. The component creates its own document and manages it.

tomasztunik commented:
I think it's worth a try. Will look into it bit later today - thanks for joining in @jsnajdr and giving a new perspective.

tomasztunik committed:
Solve the issue with srcDoc rather than document.write

tomasztunik commented:
I think this looks cleaner now 🧹 

Thanks for the review @ellatrix and @jsnajdr. Was quite a trip into the iframe land :)

jsnajdr commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -264,6 +271,8 @@ function Iframe(
 				ref={ useMergeRefs( [ ref, setRef ] ) }
 				tabIndex={ tabIndex }
+				// Correct doctype is required to enable rendering in standards mode
+				srcDoc="<!doctype html><html><head>yyy</head><body>xxx</body></html>"
```
Here you only need to create the `<html>` element, with no content. The `createPortal` call will populate the `.documentElement` (which points to the `html` element) with `head` and `body`. As is the custom, the root element where React UI is mounted should be empty.

tomasztunik replied:
Good catch - removing - and left some debug noise on top.

jsnajdr commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -174,9 +174,16 @@ function Iframe(
+			const { readyState, documentElement, compatMode } = contentDocument;
+
+			// As srcDoc loads contents asynchronously this will cause the iframe to
+			// load documents twice. We need to hook react to the correct contentDocument
```
When you're testing, does the document really load twice for you? With `srcDoc`, it should only load once.

When I run this little test app:
```js
function App() {
  const elRef = React.useRef();
  React.useEffect(() => {
    function onLoad() {
      const doc = elRef.current.contentDocument;
      console.log("load", doc.readyState, doc.compatMode);
    }
    elRef.current.addEventListener("load", onLoad);
    return () => elRef.current.removeEventListener("load", onLoad);
  }, []);

  return <iframe ref={elRef} srcDoc="<!doctype html>" />;
}
ReactDOM.render(<App />, document.getElementById("root"));
```
I get only one `load` event logged:
```
load complete CSS1Compat
```

tomasztunik replied:
Double checked by logging and breaking to inspect the iframe node and this is what we get:
![Screenshot 2022-03-08 at 10 51 42](https://user-images.githubusercontent.com/112270/157212319-8f297f2b-1265-434c-9c9c-8129476de131.jpg)

tomasztunik replied:
~I'm thinking if it wasn't because first load fires before ref is set and useEffect executes?~

tomasztunik replied:
~Or more likely it's a browser difference - Found this exploration https://stackoverflow.com/questions/10781880/dynamically-created-iframe-triggers-onload-event-twice/38459639#38459639 it seems it might be Chrome/webkit only?~

tomasztunik replied:
On Firefox I get:
![Screenshot 2022-03-08 at 11 15 01](https://user-images.githubusercontent.com/112270/157216135-9107c4bf-3290-4900-a52d-11f9035d93d6.jpg)

So a little bit different handling but still load fires twice

jsnajdr replied:
On my minimal example, I get only one load event even for Firefox: `load complete CSS1Compat`. I'll try it with the full Gutenberg editor.

> I'm thinking if it wasn't because first load fires before ref is set and useEffect executes?

Yes, the load certainly fires right after the element is added to DOM, which is before refs are set and effects executed. But the `iframe` element has the `srcDoc` attribute from the very beginning, and therefore should load only once.

tomasztunik replied:
Ok got it. First one is initial state not `onload` event. It is not triggered by the listener but by the manual call to the setDocumentIfReady handler. 

https://github.com/WordPress/gutenberg/blob/21e2d2249e8a478fd4c6b49e44c1c170e5c44b2d/packages/block-editor/src/components/iframe/index.js#L213-L220

We have the comment here that load happens only in `firefox` but now with srcDoc in place it will fire everywhere it seems.

Should we remove L213-215 and leave only the onload event? This way we could remove the comment here and remove or simplify the comment inside the `setDocumentIfReady` and maybe remove the ready check before we initialise the iframe content.

jsnajdr replied:
> Should we remove L213-215 and leave only the onload event?

Yes, it's most likely guaranteed that the load is async, and the `useRefEffect` callback is always called before the load finishes. We can rely on the `load` event to fire and avoid checking synchronously.

ellatrix replied:
Yes, probably good to only use the load event now. The reason we didn't use the load event for all browsers was because we could start loading the editor much quicker.

tomasztunik replied:
I've removed the comment and direct call and removed the `compatMode` check from ready check. I believe with how much we've simplified it the simple comment around iframe event binding and reason behind srcDoc with E2E tests explaining it a bit more and guarding for ensuring we have the right `compatMode` should be enough.

tomasztunik committed:
Clean up the iframe srcDoc

ellatrix reviewed (changes requested):
Could you change it so we're just using the load event?

tomasztunik committed:
Simplify the code based on findings

jsnajdr commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -176,7 +176,7 @@ function Iframe(
 			const { readyState, documentElement } = contentDocument;
 
-			if ( readyState !== 'interactive' && readyState !== 'complete' ) {
+			if ( readyState !== 'complete' && readyState !== 'interactive' ) {
```
In the end this part of the patch only swaps the order of the two `readyState` checks and is redundant.

jsnajdr commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -203,11 +203,7 @@ function Iframe(
+		// Document set with srcDoc is not immediately ready.
 		node.addEventListener( 'load', () => {
 			setDocumentIfReady();
 		} );
```
This can be done without the arrow function:
```js
node.addEventListener( 'load', setDocumentIfReady );
```
`setDocumentIfReady` will receive an `event` parameter which it doesn't use anyway.

Also, to tidy things up, the effect callback should return a cleanup function:
```js
return () => node.removeEventListener( 'load', setDocumentIfReady );
```

tomasztunik replied:
done and done :)

jsnajdr reviewed (approved):
Looks good! I posted two comments that are more about code quality rather than functionality. 🚢

tomasztunik committed:
cleanup + ensure we unbind the event

jsnajdr commented:
@tomasztunik Can you rebase the PR branch onto the latest `trunk`? That could make the React Native e2e tests green again.

tomasztunik commented:
@jsnajdr seems it worked 👍

jsnajdr commented:
Nice! Now only @ellatrix's :-1: blocks the merge. There used to be a "dismiss merge" action (the requested changes have been done) in the GitHub UI, but I don't see it now.

tomasztunik commented:
It is under the ellipsis next to reviewer name but if it's available depends on how the repository is setup.  

Thanks again for great review session on this to both of you! I feel like I've recapped most of my knowledge on iframes and learned some on top :)

tomasztunik commented:
Hey @ellatrix, Could you please do the final check? Appreciated!

gziolo commented:
Everything looks green on CI. I see that all feedback was addressed so let's get this one in.
